nmigen_boards/orangecrab_r0_1.py

```diff
@@ -0,0 +1,124 @@
+import os
+import subprocess
+import shutil
+
+from nmigen.build import *
+from nmigen.vendor.lattice_ecp5 import *
+from .resources import *
+
+
+__all__ = ["OrangeCrabR0_1Platform"]
+
+
+class OrangeCrabR0_1Platform(LatticeECP5Platform):
+    device      = "LFE5U-25F"
+    package     = "MG285"
+    speed       = "8"
+    default_clk = "clk"
+    resources   = [
+        Resource("clk", 0, Pins("A9", dir="i"),
+                 Clock(48e6), Attrs(IO_TYPE="LVCMOS33")),
+
+        # Used to reload FPGA configuration.
+        # Can enter USB bootloader by assigning button 0 to program.
+        Resource("program", 0, PinsN("V17", dir="o"), Attrs(IO_TYPE="LVCMOS33")),
+
+        RGBLEDResource(0,
+            r="V17", g="T17", b="J3", invert=True,
+            attrs=Attrs(IO_TYPE="LVCMOS33")),
+
+        *SPIFlashResources(0,
+            cs="U17", clk="U16", miso="T18", mosi="U18", wp="R18", hold="N18",
+            attrs=Attrs(IO_TYPE="LVCMOS33"),
+        ),
+
+        Resource("ddr3", 0,
+            Subsignal("rst",     PinsN("B1", dir="o")),
+            Subsignal("clk",     DiffPairs("J18", "K18", dir="o"), Attrs(IO_TYPE="SSTL135D_I")),
+            Subsignal("clk_en",  Pins("D6", dir="o")),
+            Subsignal("cs",      PinsN("A12", dir="o")),
+            Subsignal("we",      PinsN("B12", dir="o")),
+            Subsignal("ras",     PinsN("C12", dir="o")),
+            Subsignal("cas",     PinsN("D13", dir="o")),
+            Subsignal("a",       Pins("A4 D2 C3 C7 D3 D4 D1 B2 C1 A2 A7 C2 C4", dir="o")),
+            Subsignal("ba",      Pins("B6 B7 A6", dir="o")),
+            Subsignal("dqs",     DiffPairs("G18 H17", "B15 A16", dir="io"),
+                      Attrs(IO_TYPE="SSTL135D_I", TERMINATION="OFF",
+                      DIFFRESISTOR="100")),
+            Subsignal("dq",      Pins("F17 F16 G15 F15 J16 C18 H16 F18 C17 D15 B17 C16 A15 B13 A17 A13", dir="io"),
+                      Attrs(TERMINATION="75")),
+            Subsignal("dm",      Pins("G16 D16", dir="o")),
+            Subsignal("odt",     Pins("C13", dir="o")),
+            Attrs(IO_TYPE="SSTL135_I", SLEWRATE="FAST")
+        ),
+
+        Resource("ddr3_pseudo_power", 0,
+            # pseudo power pins, leave these at their default value
+            Subsignal("vcc_virtual", PinsN("A3 B18 C6 C15 D17 D18 K15 K16 K17", dir="o")),
+            Subsignal("gnd_virtual", Pins("L15 L16 L18", dir="o")),
+            Attrs(IO_TYPE="SSTL135_II", SLEWRATE="FAST")
+        ),
+
+        Resource("mcu", 0,
```

Comment:
What is "mcu"?

Reply:
There's an ATSAMD microcontroller on board (originally to act as an ADC for the feather spec, but it can be used for whatever) and it has 4 GPIOs for communication with the FPGA.

Reply:
That seems like it should be a connector, I think; the resource you added wouldn't be very convenient if the MCU actually presented some sort of fixed function on those pins, with fixed directions.

Reply:
Yeah that makes more sense, thanks.